bdk_electrum/src/lib.rs

```diff
@@ -0,0 +1,608 @@
+
+        let res_height = match client
+            .script_get_history(spk)?
+            .into_iter()
+            .find(|r| r.tx_hash == txid)
+        {
+            Some(r) => r.height,
+            None => continue,
+        };
+
+        let tx_height = match res_height {
+            // Electrum server API specifies that height <= 0 is considered unconfirmed
+            h if h <= 0 => {
+                debug_assert!(
+                    h == 0 || h == -1,
+                    "unexpected height ({}) from electrum server",
+                    h
+                );
+                TxHeight::Unconfirmed
+            }
+            h => {
+                let h = h as u32;
+                if h > tip.height {
+                    TxHeight::Unconfirmed
+                } else {
+                    TxHeight::Confirmed(h)
+                }
+            }
+        };
+
+        if let Err(failure) = update.insert_tx(txid, tx_height) {
+            match failure {
+                sparse_chain::InsertTxError::TxTooHigh { .. } => {
+                    unreachable!("we should never encounter this as we ensured height <= tip");
+                }
+                sparse_chain::InsertTxError::TxMovedUnexpectedly { .. } => {
+                    return Err(InternalError::Reorg);
+                }
+            }
+        }
+    }
+    Ok(())
+}
+
+/// Populate an update [`SparseChain`] with transactions (and associated block positions) from
+/// the transaction history of the provided `spks`.
+fn populate_with_spks<K, I, S>(
+    client: &Client,
+    update: &mut SparseChain,
+    spks: &mut S,
+    stop_gap: usize,
+    batch_size: usize,
+) -> Result<BTreeMap<I, (Script, bool)>, InternalError>
+where
+    K: Ord + Clone,
+    I: Ord + Clone,
+    S: Iterator<Item = (I, Script)>,
+{
+    let tip = update.latest_checkpoint().map_or(0, |cp| cp.height);
+    let mut unused_spk_count = 0_usize;
+    let mut scanned_spks = BTreeMap::new();
+
+    loop {
+        let spks = (0..batch_size)
+            .map_while(|_| spks.next())
+            .collect::<Vec<_>>();
+        if spks.is_empty() {
+            return Ok(scanned_spks);
+        }
+
+        let spk_histories = client.batch_script_get_history(spks.iter().map(|(_, s)| s))?;
+
+        for ((spk_index, spk), spk_history) in spks.into_iter().zip(spk_histories) {
+            if spk_history.is_empty() {
+                scanned_spks.insert(spk_index, (spk, false));
+                unused_spk_count += 1;
+                if unused_spk_count > stop_gap {
+                    return Ok(scanned_spks);
+                }
+                continue;
+            } else {
+                scanned_spks.insert(spk_index, (spk, true));
+                unused_spk_count = 0;
+            }
+
+            for tx in spk_history {
+                let tx_height = match tx.height {
+                    // Electrum server API specifies that height <= 0 is considered unconfirmed
```

Comment:
So the genesis block tx is literally impossible to represent correctly :facepalm: ? It sounds like we have to special case the tx to check if it's the genesis tx so it doesn't show that this tx is always unconfirmed (any by extension all descendants are unconfirmed).

Reply:
On it now.